packages/teraslice-state-storage/src/elasticsearch-state-storage/index.ts

```diff
@@ -0,0 +1,197 @@
+
+import { DataEntity, Logger, TSError } from '@terascope/job-components';
+import esApi, { Client } from '@terascope/elasticsearch-api';
+import { Promise as bPromise } from 'bluebird';
+import { ESStateStorageConfig, ESBulkQuery, ESQUery, MGetResponse } from '../interfaces';
+import CachedStateStorage from '../cached-state-storage';
+
+export default class ESCachedStateStorage {
+    private index: string;
+    private type: string;
+    private IDField: string;
+    private concurrency: number;
+    private sourceFields: string[];
+    private chunkSize: number;
+    private persist: boolean;
+    private persistField: string;
+    private es: Client;
+    private cache: CachedStateStorage;
+
+    constructor(client: Client, logger: Logger, config: ESStateStorageConfig) {
+        this.index = config.index;
+        this.type = config.type;
+        this.IDField = config.id_field;
+        this.concurrency = config.concurrency;
+        this.sourceFields = config.source_fields;
+        this.chunkSize = config.chunk_size;
+        this.persist = config.persist;
+        this.persistField = config.persist_field || config.id_field;
+        this.cache = new CachedStateStorage(config);
+        this.es = esApi(client, logger);
+    }
+
+    private getIdentifier(doc: DataEntity) {
+        const id =  doc.getMetadata(this.IDField);
+        if (id == null) throw new TSError(`There is no field "${this.IDField}" set in the metadata for doc: ${JSON.stringify(doc)}`);
```

Comment:
It would be good to catch an empty string as well.

```suggestion
        if (!id) throw new TSError(`There is no field "${this.IDField}" set in the metadata`, {
             context: { doc }
        });
```